packages/origin-247-certificate/src/offchain-certificate/certificate.aggregate.ts

```diff
@@ -3,65 +3,87 @@ import {
     CertificateClaimPersistedEvent,
     CertificateEventType,
     CertificateIssuancePersistedEvent,
+    CertificateIssuedEvent,
+    CertificatePersistErrorEvent,
     CertificateTransferPersistedEvent,
     CertificateTransferredEvent,
     ICertificateEvent
 } from './events/Certificate.events';
-import { IIssueCommand } from '../types';
 import { CertificateErrors } from './errors';
 import { BigNumber } from 'ethers';
 import { compareDates } from '../utils/date-utils';
 import { ICertificateReadModel } from '..';
 
+type Cert<T> = ICertificateReadModel<T>;
+
 export class CertificateAggregate<T> {
-    private certificate: ICertificateReadModel<T> | null = null;
+    private certificate: Cert<T>;
 
     private constructor(events: ICertificateEvent[]) {
-        events
-            .sort((first, second) => compareDates(first.createdAt, second.createdAt))
-            .forEach((event) => this.apply(event));
+        const sortedEvents = [...events].sort((a, b) => compareDates(a.createdAt, b.createdAt));
+        const [firstEvent, ...restEvents] = sortedEvents;
+
+        if (!firstEvent) {
+            throw new CertificateErrors.CertificateNoEvents();
+        }
+
+        if (firstEvent.type !== CertificateEventType.Issued) {
+            throw new CertificateErrors.CertificateNotIssuanceEvent(
+                firstEvent.internalCertificateId,
+                firstEvent.type
+            );
+        }
+
+        const initialCertificate = this.createCertificate(firstEvent as CertificateIssuedEvent<T>);
+
+        this.certificate = {
+            ...restEvents.reduce(this.apply.bind(this), initialCertificate),
+            isSynced: this.reduceIsSynced(events)
+        };
     }
 
-    public apply(event: ICertificateEvent): void {
+    private apply(certificate: Cert<T>, event: ICertificateEvent): Cert<T> {
```

Comment:
ensuring that `apply` returns  something allows TS to properly ensure, that everything is handled in `switch`